[RFC] Template prototype using jinja2 templating engine

evenl opened the pull request:
Sorry for this, my bad, I had to create a new pull request:
This is an continuation of https://github.com/zephyrproject-rtos/zephyr/pull/10873

**Goal**

To propose the jinja2 template engine as an alternative to the COG
based solution proposed for code generation.

**Rationale**

The version of COG proposed to be used in Zephyr is a highly modified
version of the original COG implementation. Which means that we can
not leverage future improvements made to COG or contribute our own
modifications back to COG.

This commit https://github.com/evenl/zephyr/commit/d8c986c9bfc2f207adc31d7e16b91e32e753ea92
shows what code can be removed by using a jinja instead of a COG base solution

The Jinja2 engine however can be used as a pip package dependency
without modifications.

COG is more or less straight forward python which means that a generator
must more or less be written for each case. Which means that even the
simplest things will need a lot of boiler code just to output something.

**Example:**

Data in is “dev_name”

Cog:
```
/**
 * @code{.codegen}
 *
 *     codegen.outl(“#ifndef {}_INCLUDE_H”.format(dev_name))
 *     codegen.outl(“#define {}_INCLUDE_H”.format(dev_name))
 *
 * @endcode{.codegen}
 */
```

Jinja2:
```
    #ifndef {{ dev_name }}_INCLUDE_H
    #define {{ dev_name }}_INCLUDE_H
```

**TODO:**

The Jinja2 implementation is on par with what is currently implemented using the COG
solution. But because the COG solution give the user the ability to directly write
python code directly in the template there may be a need to develop some custom filters/
helper functions for the jinja2 solution.

This code is based on this code https://github.com/galak/zephyr/tree/EDTS which is not 
yet in master. 

**Other advantages:**

        * Big active community
        * Standardized and documented (http://jinja.pocoo.org)
        * Has been in development for a long time with multiple stable releases
        * Extensive enough to be a one-stop-shop for all code generation in the Zephyr project

b0661 committed:
scripts: extract_dts_includes.py: extract reg to EDTS database

Extend reg directive handling to fill the EDTS datanbase.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

b0661 committed:
scripts: extract_dts_includes.py: extract clocks to EDTS database

Extend clocks directive handling to fill the EDTS database.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 committed:
scripts: extract_dts_includes.py: default extract to EDTS database

Extend default directive handling to fill the EDTS database.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 committed:
scripts: extract_dts_includes.py: create EDTS database for codegen

Create extended DTS database to be used by codegen and others.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

erwango committed:
scripts: extract_dts_includes.py: extract compatible to EDTS database

Get compatible info available by EDTS.
This way the compatible info of all activated devices can be used in
drivers.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

erwango committed:
scripts: edts: Fill-in 'controllers' views

Fill-in controllers hierarchy in edts database.
Aim is to provide a per controller complete hierarchized view
of the various controller types.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scrpits/dts: edts: add alias treatment for edts generation

Add an 'aliases' dict within generated edts.json and
fill in alias information when required in device-struct.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scripts/dts: edts: Get children at their parent's

Nodes without compatible are children of ancestor node.
We want to inject these node in a children dict of parent node.
In order to detect them more easily split get_compat function
into get_node_compat and get_parent_compat.
Then, function generate_node_definitions in extract_dts_includes
script is modified to parse parents first then children so parents
are injected in edts first and ready to welcome their children.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scripts/dts: edts: insert chosen to edts database

Add chosen structure ti edts database

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak committed:
scripts: extract_dts_includes: extract gpios to EDTS database

Extract gpio property for injection in EDTS.
Cleanup no more needed function from extract_dts_includes script

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

erwango committed:
scripts/dts: edts: insert 'status' node property

Add 'status' in edts devices struct when available.
This operation is done as part of default class and called
before usual properties parsing since 'status' property
doesn't have binding.
Default class is modified in order not to generate #define
if def_label is set to None (hence limited to edts injection).

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts: bindings: Add basic bindings dedicated to dts validation

Add 'refsoc' yaml dts bindings.
These bindings will be used in a refboard dts file which will
be the reference for dts/edts sanity checks.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: Add refboard dummy board

Aim of this board is to provide a reference device tree that
will be used for dts/edts scripts sanity check.
While in theory this board does not use ARM architecture and should
have its own 'ref' arch, it is set under arm/ right now in order
to avoid providing a dedicated dummy arch.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scripts/dts: Introduce edts.json generation

Introduce python class for extended device tree database generation
and extraction.
Extended device tree database is a json formatted database.
It contains python like dict representation of device tree information
amended with zephyr yaml bindings.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>
Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts/bindings: Add yaml field 'type' for generic zephyr type

Add new 'type' field in dts yaml bindings. It is meant to
provide zephyr generic device type (i2c, spi, ...).
This information could be used for various operations that
require to have information on the generic type of a device
or a compatible (CI triage, board documentation, ...)

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scripts: extract_dts_includes.py: extract heuristics to EDTS database

Extend directive handling to use heuristics to fill the EDTS database.

heuristics was called when treating 'compatible' property
on the assumption it is mandatory.
Actually some nodes (children) don't have 'compatible', so
move heuristics out or properties parsing and process them
just before node properties treatment.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>
Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scripts/edts: inject device-type in edts database.

Extract and inject zephyr generic device-type in edts database.
This information is extracted from dts yaml bindings 'type' field.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scripts/dts: edts: generate a unique id for each device

Generate a unique_id with the following format for each device:
<PARENT>_<COMPAT_PARENT>_<ADDRESS_NODE>_<COMPAT_NODE_ADDRESS>

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak committed:
simple tests

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

compat edts test

compat.dts

Add test for compat checking

galak committed:
REG test

evenl committed:
Switched from COG to jinja2 based code generation. Ported instantiation
code generation for the I2C and SPI driver for ST devices use jinja2

Signed-off-by: Even Falch-Larsen even.falch-larsen@nordicsemi.no
(cherry picked from commit 843cfad25ea6c1c6347b77365f8e4d5e171c868d)

galak committed:
edtsdevice: Add Regs upport

galak committed:
Add unique id

galak committed:
scripts: edts: Add api to get device_ids based on compat

Added get_device_ids_by_compatible to EDTSConsumerMixin

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: edts: Add api to get device_id based on label

Added get_device_id_by_label to EDTSConsumerMixin

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: edts: Add api to get list of compatiables

Added get_compatibles to EDTSConsumerMixin

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: edts: Add api to get list of aliases

Added get_aliases to EDTSConsumerMixin

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: edts: Add api to get list of chosen

Added get_chosen to EDTSConsumerMixin

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: edts: Add api to get list of device types

Added get_device_types to EDTSConsumerMixin

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: edts: Add api to get list of controllers

Added get_controllers to EDTSConsumerMixin

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
test

erwango committed:
scripts/dts: provide edts sanity check scripts

Validate generated etds.json against a reference edts file.
Script first checks edts structure then content of various elements
and finally devices parameters values.
This last step is done using a specific method for ref_edts
parameter extraction so we can verify behavior of
edtsdatabase class extraction function.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

evenl committed:
Modifying codegen documentation

Signed-off-by: Even Falch-Larsen even.falch-larsen@nordicsemi.no
(cherry picked from commit d74134ea0ee483c98fad82fef1b383afbcf3da2c)

galak committed:
edts: reg testing - hack

galak committed:
scripts: extract_dts_includes.py: extract interrupts to EDTS database

Extend interrupts directive handling to fill the EDTS database.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
Add EDTSDevice irq support

galak committed:
test edtsdevice irq

evenl committed:
Updated prototype to generate a .h file which is included in a .c file

Signed-off-by: Even Falch-Larsen even.falch-larsen@nordicsemi.no
(cherry picked from commit 38d93bc53ea978ccb833bbf7dd29e12286eccb29)

evenl committed:
Template updates

* Can now set and get parameters instead of direcly
  populating a dictionary
* IRQ_CONNECT and irq_enable is now autogenerated
* Removed the need for setting CODEGEN_OUTPUT if only the
  last extention of the template filename should be removed

Signed-off-by: Even Falch-Larsen even.falch-larsen@nordicsemi.no
(cherry picked from commit 414e44724fccefc7af3e4f8549690b71d58ec503)

evenl committed:
Exposing EDTS interfaces

Exposing EDTSDatabase and EDTSDevice API for templates
through the edts_api data object

Signed-off-by: Even Falch-Larsen even.falch-larsen@nordicsemi.no

evenl committed:
Moved generation parameters

Moved generation parameters into a dictonary in the main
template to removed the need for set/get_param macros

Signed-off-by: Even Falch-Larsen even.falch-larsen@nordicsemi.no

evenl commented:
Yes, sorry for that whole thing, it was late, I was not sure what I was doing ;)

galak commented:
What board do you try this out on?

When I try to build disco_l475_iot1 I get:

```
[galak@spiff build]$ cmake -DBOARD=disco_l475_iot1  ..

...
-- Will generate for target drivers__spi
--  from '/home/galak/git/zephyr/drivers/spi/spi_ll_stm32_device_template.h'
--  to   '/home/galak/git/zephyr/samples/hello_world/build/zephyr/drivers/spi/spi_ll_stm32_device_template'
-- Configuring done
CMake Error in ../../drivers/spi/CMakeLists.txt:
  Cannot find source file:

    /home/galak/git/zephyr/drivers/spi/spi_ll_stm32_device_template.h

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
  .hpp .hxx .in .txx


CMake Error: Cannot determine link language for target "drivers__spi".
CMake Error: CMake can not determine linker language for target: drivers__spi
-- Generating done
-- Build files have been written to: /home/galak/git/zephyr/samples/hello_world/build
```

evenl commented:
I have not kept the SPI driver up to date with the latest changes, only the I2C driver is currently working. I'll update the SPI driver when I am done with the current update.

evenl committed:
Obsoleted the meta file

Removed the need for the metafile for injecting metadata,
everything is handled in one template now

Signed-off-by: Even Falch-Larsen even.falch-larsen@nordicsemi.no

evenl committed:
Renamed tool to JinjaGen

Renamed the tool to JinjaGen to diversify it from codegen

Signed-off-by: Even Falch-Larsen even.falch-larsen@nordicsemi.no

evenl commented:
> ooks like the ISR logic needs some improvement. The existing code looks like:

That one was working before, something probably happen when I changed to the EDTS api, I'll fix it.
Yes, it support the ifdef logic, it will render the correct IRQ_CONNECT lines based on the information in the DTS data. I think it is just reading the wrong field for some reason.

SebastianBoe commented on cmake/extensions.cmake:
```diff
@@ -344,6 +344,17 @@ function(generate_inc_file_for_gen_target
+    string(REPLACE "." "_" basename ${basename})
+    string(REPLACE "@" "_" basename ${basename})
+    set(generated_target_name "gen_${basename}_${random_chars}" PARENT_SCOPE)
+endfunction()
```
Code like this should be upstreamed early to minimize the size of the PR.

https://github.com/zephyrproject-rtos/zephyr/pull/10969

evenl commented on templates/device_declare.jinja2:
```diff
@@ -0,0 +1,98 @@
+	IRQ_CONNECT({{ device.get_irq(irq) }}, \
+		    {{ device.get_irq_prop(irq, 'priority') }}, \
+		    {{ params.irq_func }}_{{ irq }}, \
+		    DEVICE_GET({{ gen_device_name(device) }}, \
```
fixed

galak commented:
> That one was working before, something probably happen when I changed to the EDTS api, I'll fix it.
> Yes, it support the ifdef logic, it will render the correct IRQ_CONNECT lines based on the information in the DTS data. I think it is just reading the wrong field for some reason.

We probably need some enhancements to the EDTS Device API around IRQs.  We're going to need more logic in gen_irq_connect macro to handle the various cases:

1. There are no IRQs in the dts but we pass IRQ params to the device macro
2. How do we associate IRQ function name w/IRQs (we need to support both by DTS index & DTS interrupt-names)
3. How do we handle cases like shared IRQ - We should be able to tell by the # of IRQs or name in the driver specific template

evenl commented on drivers/i2c/CMakeLists.txt:
```diff
@@ -18,8 +18,12 @@ zephyr_library_sources_ifdef(CONFIG_I2C_SBCON		i2c_sbcon.c)
+zephyr_library_sources_ifdef(CONFIG_I2C_STM32_V1        i2c_ll_stm32.c)
+zephyr_library_sources_ifdef(CONFIG_I2C_STM32_V2        i2c_ll_stm32.c)
+
+zephyr_library_sources_jinjagen_ifdef(CONFIG_I2C_STM32_V1 i2c_ll_stm32_device.h.jinja2 SEARCH_PATH ${PROJECT_SOURCE_DIR}"/templates/")
```
I can drop the SEARCH_PATH and make it hard coded to comply with the requirements, but I think it a good feature to driver developers point to places where they store commonly used templates.

galak replied:
I think we should have the ${PROJECT_SOURCE_DIR}"/templates/" should be included by default since that will be common templates we define for driver developers.

SebastianBoe replied:
+1

Also, it should be ${ZEPHYR_BASE}/templates

evenl replied:
Ok, ${ZEPHYR_BASE}/templates is added by default, any other folder can still be added through the same mechanism.

SebastianBoe commented on cmake/extensions.cmake:
```diff
@@ -353,22 +364,141 @@ function(generate_inc_file_for_target
-    RANDOM
-    LENGTH 8
-    random_chars
+function(target_sources_jinjagen
```
This API must be documented.

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+Jinja2 files are rendered into other files using the cmake extensions functions
+
+  * zephyr_sources_codegen(..)
+  * zephyr_library_sources_codegen(..). 
```
jinjagen

evenl replied:
ok

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+
+The generated source files are added to the Zephyr sources. During build the 
+jinja2 files are processed by jinja2 template engine and the generated source
+files are written to the CMake binary directory.
```
s/CMake binary directory/build directory/

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+  * zephyr_sources_codegen(..)
+  * zephyr_library_sources_codegen(..). 
+
+The generated source files are added to the Zephyr sources. During build the 
```
"During build" doesn't sound right.

dbkinder replied:
``When the project is built, the ...``

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+
+A jinja2 template is written in a domain specific language, but inherited many
+consepts from python, which means that anyone familiar with python should
+learn to write jinja2 templates with ease. 
```
A jinja2 template is written in a domain specific language. The language inherits
concepts from python, so anyone familiar with python should find it easy to learn
how to write jinja2 templates.

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+===========
+
+All templates have access to a data object called ``data``. Through this data
+object will the template have access to information like all the elements in
```
"object will the template"

object the template will

SebastianBoe replied:
"to information like all"

to information, like

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+	{% endfor %}
+
+
+and runtime information like defines that has been injected using the CODEGEN_DEFINES
```
JINJAGEN?

dbkinder replied:
change ``like`` to ``such as``

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+	{% endfor %}
+
+
+and runtime information like defines that has been injected using the CODEGEN_DEFINES
```
"runtime information" could be confusing. I see that it was used in the template flow picture as well.

"and configure time information", or just "and information", would avoid this confusion.

evenl replied:
"code generation time" or "template processing time" maybe? The point is that this information can be insert at the time a template is processed, which means that the same template can be reused with different "runtime" information to generate different output. This can be a very usefull feature, but I see that "runtime" can be confusing, at least as a term in the documentation.

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+
+.. code-block:: c
+
+	{{ data['runtime']['defines']['MY_DEFINE'] }}
```
W need to rename "runtime" to something else.

"cmake" maybe? I'm not yet familiar with what it is.

But the term runtime should always be referencing the actual fw running.

evenl replied:
Runtime refer to the runtime of the code generator script, while cmake would refer the runtime of cmake which may spawn multiple code generation sessions with different code generation runtime data. Maybe straight forward "jinjagen"?

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+Built-in
+--------
+
+Jinja2 has a set of built in filter that in many cases will be familiar to anyone who
```
built-in filters

evenl replied:
ok

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+
+.. code-block:: c
+
+        {{ "fOOBAR!"|capitalize() }} => Foobar!
```
How come you used \| in the previous example, but not now?

evenl replied:
Typo maybe? I am not sure what you referring to? Did it say 
`{{ "fOOBAR!"\|capitalize() }} => Foobar!` before? That would not have worked.

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+------
+
+Filters can be added to the jinja2 engine by defining custom filters in the python
+script that call the jinja2 engine. This way can functionality easily be added to 
```
that calls the

evenl replied:
ok

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+------
+
+Filters can be added to the jinja2 engine by defining custom filters in the python
+script that call the jinja2 engine. This way can functionality easily be added to 
```
way functionality can

evenl replied:
ok

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+
+Filters can be added to the jinja2 engine by defining custom filters in the python
+script that call the jinja2 engine. This way can functionality easily be added to 
+the jinja2 engine to tailer it for any particular application.
```
tailor

evenl replied:
ok

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+script that call the jinja2 engine. This way can functionality easily be added to 
+the jinja2 engine to tailer it for any particular application.
+
+There is no square root filter built into jinja, this can be added by using 
```
jinja, but this can

evenl replied:
ok

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+        def squarerootfilter(value):
+                return sqrt(value)
+
+        environment.filters['sqrt'] = squarerootfilter
```
Could we include here what file a user would modify to achieve this, and perhaps a disclaimer about whether it is supported to have out-of-tree filters or not?

(Unless this is documented elsewhere that I haven't seen)

evenl replied:
Ok, currently you add the custom filters to the main jinjagen python script in the scripts folder.

Currently out-of-tree filters are not supported, technically it is just a few simple lines of python, but conceptually it will add complexity because we will need to define some way (metadata, mechanism) to insert the out-of-tree filters and also add a version scheme for the tool it self and the filters. 

But I can add simple support for it in the prototype if you want to see how it looks/works.

evenl replied:
Wait until the RFC process is done

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+===============================
+
+Templates can extend and/or reuse each other through multiple mechanisms. To be able
+extend or reuse a template, the template must be stored in one of the folders listed
```
Too much repetition.

"A house can be painted or washed through multiple mechanisms. To paint or wash a house, "

Maybe drop the first sentence?

evenl replied:
ok

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+Templates can extend and/or reuse each other through multiple mechanisms. To be able
+extend or reuse a template, the template must be stored in one of the folders listed
+in the search path list. Search paths can be set using the SEARCH_PATH key word in the
+CMakeLists.txt file.
```
How is the "search path list" defined and modified?

perhaps we should be saying ....

> the template must be stored in one of the folders listed in the CMake variable 'SEARCH_PATH'.

evenl replied:
Ok, rephrased that section and added an example.

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+Include
+-------
+
+Include works more or less like include pre-processor directive in C, and will
```
more or less like the \#include

evenl replied:
ok

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+
+	{% import 'template_with_macros.jinja2' as macros %}
+
+	{{ macros.an_imported_macro() }}
```
Do we need to document how Jinja2 works? Isn't that done best by Jinja2?

Note that we do not document how yaml, DeviceTree, Kconfig, C, or the CMake language behaves.

I understand we might be doing so because Codegen did so, but with Jinja2 this is different because Codegen forked the language and was forced to also fork the documentation.

evenl replied:
Well, it's already there so let's keep it for now, at least through the RFC process.

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+====================================
+
+Code generation has to be invoked as part of the build process. Zephyr uses
+`CMake <https://cmake.org/>`_ as the tool to manage building the project.
```
> Zephyr uses `CMake <https://cmake.org/>`_ as the tool to manage building the project.

This sentence can be dropped as it is covered elsewhere.

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+Code Generation in the Build Process
+====================================
+
+Code generation has to be invoked as part of the build process. Zephyr uses
```
"has to be invoked"

Code generation is invoked as part

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+`CMake <https://cmake.org/>`_ as the tool to manage building the project.
+
+A template file that must be rendered before it can be used in a project 
+by one of the following commands in a `CMakeList.txt` file.
```
A template file can be inserted into the build by one of the following CMake functions.

evenl replied:
ok

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+
+.. code-block:: c
+
+    zephyr_sources_codegen(codegen_file.c [CODEGEN_DEFINES defines..] [SEARCH_PATH target.. path..])
```
jinja

evenl replied:
ok

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+
+    zephyr_library_sources_codegen(codegen_file.c [CODEGEN_DEFINES defines..] [SEARCH_PATH target.. path..])
+
+    zephyr_library_sources_codegen_ifdef(ifguard codegen_file.c [CODEGEN_DEFINES defines..] [SEARCH_PATH target.. path..])
```
You can drop the _ifdef functions from the documentation.

evenl replied:
ok

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+
+The arguments given by the ``CODEGEN_DEFINES`` keyword have to be of the form
+``define_name=define_value``. The arguments becomes a part of the data object
+that can be accessed by any template and can be accessed through ``data['runtime']['defines'][define_name]``
```
a part of the data object and can be accessed through

evenl replied:
ok

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+
+    zephyr_sources_codegen(codegen_file.c [CODEGEN_DEFINES defines..] [SEARCH_PATH target.. path..])
+
+    zephyr_sources_codegen_ifdef(ifguard codegen_file.c [CODEGEN_DEFINES defines..] [SEARCH_PATH target.. path..])
```
"target.. path.."

?

path1 [path2] [...]

?

evenl replied:
yes, fixed

SebastianBoe commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+
+The arguments given by the ``CODEGEN_DEFINES`` keyword have to be of the form
+``define_name=define_value``. The arguments becomes a part of the data object
+that can be accessed by any template and can be accessed through ``data['runtime']['defines'][define_name]``
```
Can you elaborate on when you would want to use a CODEGEN_DEFINE?

I am concerned we have too many mechanisms with overlapping functionality and would like to get rid of it if possible.

evenl replied:
So more or less the same reasons for using gcc -D, but for a code generation session. 

* it can be a feature toggle, 
* render multiple variants of the same template
   * Similar hardware module share the same driver, but one instance need some errata code

on the top of my head

SebastianBoe commented on cmake/extensions.cmake:
```diff
@@ -353,22 +364,141 @@ function(generate_inc_file_for_target
+  string(REGEX REPLACE "([^;]+)" "-I;\\1"
+	  JINJAGEN_SEARCH_PATH "${JINJAGEN_SEARCH_PATH}")
+
+  # Get all the files that make up codegen for dependency
```
for dependency reasons

evenl replied:
ok

SebastianBoe reviewed (dismissed):
From a build perspective I approve of the concepts proposed in this RFC.

There is still some work to do, and I hope that this RFC can
be approved in reasonable time so we can turn this into a PR
and get to work.

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+Import
+------
+
+The import keyword will import all macros defined in another template into a namespace.
```
Join the two sentences:

    The import keyword will import all macros defined in another template into a namespace,
    similar to how import in python works.

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+Extends
+-------
+
+The extends keyword is used to extend and overload macros from another template. 
```
Join these sentences together too (and fix a typo):

    The extends keyword is used to extend and overload macros from another template,
    similar to how a child class extends a parent class in object oriented programming.

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+.. _codegen_intro:
+
+Introduction
+============
```
Our project convention is for headings to start with a ``###`` underline (for the title), followed by ``***``, ``===``, and ``---`` for nested levels.

evenl replied:
I was looking for some documentation for this, but ended up just following the Sphinx documentation. I'll update the topics

evenl replied:
Ok

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+learn to write jinja2 templates with ease. 
+
+Data object
+===========
```
Use ``****``'s for underline here

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+	{{ data['runtime']['defines']['MY_DEFINE'] }}
+
+Filters
+=======
```
Use ``****``'s for underline here

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+        {{ "fOOBAR!"|capitalize() }} => Foobar!
+
+Custom
+------
```
Use ``====``'s for underline here

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+
+
+Extending and reusing templates
+===============================
```
Use ``****``'s for underline here

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+CMakeLists.txt file.
+
+Include
+-------
```
Use ``====``'s for underline here

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+
+
+Import
+------
```
Use ``====``'s for underline here

evenl replied:
Ok

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+
+
+Extends
+-------
```
Use ``====``'s for underline here

evenl replied:
Ok

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+	{% endblock %}
+
+Block
+^^^^^
```
Use ``====``'s for underline here  (this is a sibling section to the previous heading, right?)

evenl replied:
ok

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+    {% endblock %}
+
+Code Generation in the Build Process
+====================================
```
Use ``****``'s for underline here

evenl replied:
Ok

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,223 @@
+``define_name=define_value``. The arguments becomes a part of the data object
+that can be accessed by any template and can be accessed through ``data['runtime']['defines'][define_name]``
+
+The template engine can be told to look for templates in folders using the ``SEARCH_PATH`` keyword
```
The template engine looks for templates in the current folder, and in
    folders listed by the ``SEARCH_PATH`` keyword.

evenl replied:
ok

dbkinder reviewed (changes requested):
I'll take another look after the other doc comments have been added...

galak commented:
> From a build perspective I approve of the concepts proposed in this RFC.
> 
> There is still some work to do, and I hope that this RFC can
> be approved in reasonable time so we can turn this into a PR
> and get to work.

Curious about doing the template rendering at cmake cfg time (ie cmake -DBOARD=foo) vs make/ninja time.  @carlescufi mentioned that if we do it at make/ninja time we get the benefit of parallelism.  However, curious if we should be aiming to invoke python once for all the files that need to get generated.  Also, what's the build dependency if we touch a template file like between the two options?

SebastianBoe commented:
> However, curious if we should be aiming to invoke python once for all the files that need to get generated.

We should not be aiming for this.

> Also, what's the build dependency if we touch a template file like between the two options?

Doing this at Configure-time would mean that we would need to re-run CMake when a template is touched. Whereas doing this at build-time would mean that we would need to re-run the specific generator and re-compile any C files that depend on the generated file.

So doing this at configure-time would be significantly slower.

evenl commented:
I think I have resolved all comments for now

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,256 @@
+	{% endfor %}
+
+
+and runtime information such as defines that has been injected using the JINJAGEN_DEFINES
```
s/has/have/

evenl replied:
ok

dbkinder commented:
Documentation guidelines are here: https://docs.zephyrproject.org/latest/contribute/doc-guidelines.html

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,256 @@
+
+Jinja2 files are rendered into other files using the cmake extensions functions
+
+  * zephyr_sources_jinjagen(..)
```
renders better if you don't indent these lines (if you view this file in GitHub you'll see what I mean).  With too much indentation like this, the renderer adds a vertical bar on the left thinking it's a blockquote.

evenl replied:
ok

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,256 @@
+
+devicetree data,
+ 
+ .. code-block:: c
```
don't indent this directive (same blockquote rendering problem)

evenl replied:
ok

dbkinder commented on doc/subsystems/jinjagen/jinjagen.rst:
```diff
@@ -0,0 +1,256 @@
+    zephyr_library_sources_jinjagen(jinjagen_file.c [JINJAGEN_DEFINES defines..] [SEARCH_PATH path1 [path2] [..]])
+
+The arguments given by the ``JINJAGEN_DEFINES`` keyword have to be of the form
+``define_name=define_value``. The arguments becomes a part of the data object and can be accessed through 
```
s/becomes/become/

evenl replied:
ok

dbkinder reviewed (changes requested):
almost there on the docs...

dbkinder reviewed (changes requested):
doc is looking better

erwango commented on drivers/i2c/CMakeLists.txt:
```diff
@@ -16,15 +16,14 @@ zephyr_library_sources_ifdef(CONFIG_I2C_SAM_TWI		i2c_sam_twi.c)
+zephyr_library_sources_ifdef(CONFIG_I2C_STM32_V2        i2c_ll_stm32.c)
+
+zephyr_library_sources_jinjagen_ifdef(CONFIG_I2C_STM32_V1 i2c_ll_stm32_device.h.jinja2)
+zephyr_library_sources_jinjagen_ifdef(CONFIG_I2C_STM32_V2 i2c_ll_stm32_device.h.jinja2)
```
Are these last 2 lines really mandatory?

Can't we just call zephyr_library_sources_jinjagen_ifdef with the .c file of the driver we want to apply generation on?
So here would be:
```
zephyr_library_sources_jinjagen_ifdef(CONFIG_I2C_STM32_V1        i2c_ll_stm32.c)
zephyr_library_sources_jinjagen_ifdef(CONFIG_I2C_STM32_V2        i2c_ll_stm32.c)
```
So that developer only needs to care about its driver filename and that's it. Once we know the name of the file on which we want to apply device generation, things are pretty mechanical. We can build up the device filename template 'i2c_ll_stm32_device.h.jinja2'.
Makes things a bit clearer and minimize potential errors, you can't mix up between
- Use zephyr_library_sources_jinjagen_ifdef for devices generation, or
- Use zephyr_library_sources_ifdef

Besides, it is also a way to impose the format and limit the possible names variations that will pop up if we let freedom to the users. Also, having uniform names should also be helpful for various tools that we might want to use on them (I have no idea what tools yet, but we always end up using tools).

SebastianBoe replied:
Yes. Your proposal can be implemented.

If the RFC is accepted we'll get this done.

erwango commented on drivers/spi/CMakeLists.txt:
```diff
@@ -2,7 +2,7 @@ zephyr_library()
 zephyr_library_sources_ifdef(CONFIG_SPI_DW		spi_dw.c)
 zephyr_library_sources_ifdef(CONFIG_SPI_INTEL		spi_intel.c)
-zephyr_library_sources_ifdef(CONFIG_SPI_STM32		spi_ll_stm32.c)
+zephyr_library_sources_jinjagen_ifdef(CONFIG_SPI_STM32	spi_ll_stm32_device.h.jinja2)
```
I guess one line is missing here.
But, as commented above, I'll be glad with :-):
```
zephyr_library_sources_jinjagen_ifdef(CONFIG_SPI_STM32	spi_ll_stm32.c)
```

erwango commented on drivers/i2c/i2c_ll_stm32_device.h.jinja2:
```diff
@@ -0,0 +1,23 @@
+                 'init_function'      : 'i2c_stm32_init',
+		 'config_struct_body' :  config_struct_body} %}
+
+{{ i2c_stm32.device(data, ['st,stm32-i2c-v1', 'st,stm32-i2c-v2'], params) }}
```
One question on particular multi compatible support.

As you could see this file is compatible with 2 STM32 I2C IP versions. Particularly, both versions use the same code for devices.
Then, what if they used slightly different code for devices V1 and devices V2.
For instance: V1 and V2 uses same data_struct_body but have different config_struct_body

How would you support that if I still want to have device instantiation all in once place? (Assumption is only one compatible would be used for a specific board, either V1 or V2, but for all types of boards, I want to use this same file.)
This is doable with codegen since we provide 'compatible' in API, so we only have to define data_struct_body,   config_struct_body_v1 and config_struct_body_v2, then calling API twice:
-first  with devicedeclare('compat_v1', 'config_struct_body_v1', 'data_struct_body'), 
-then with devicedeclare('compat_v2', 'config_struct_body_v2', 'data_struct_body')

evenl replied:
> This is doable with codegen since we provide 'compatible' in API, so we only have to define data_struct_body, 
> config_struct_body_v1 and config_struct_body_v2, then calling API twice:
> -first with devicedeclare('compat_v1', 'config_struct_body_v1', 'data_struct_body'),
> -then with devicedeclare('compat_v2', 'config_struct_body_v2', 'data_struct_body')

That is not entirely true, there is a lot more parameters.

```
devicedeclare.device_declare( \
 *     ['st,stm32-exti', ],
 *     'CONFIG_KERNEL_INIT_PRIORITY_DEVICE',
 *     'PRE_KERNEL_1',
 *     None,
 *     'stm32_exti_init',
 *     None,
 *     device_data,
 *     None)
```
So what I did in my jinja2 solution is to move the parameters out of the function call and into an object, I know it's personal taste, but I think it is cleaner. But your use-case is totally supported, it's just a matter of creating another parameter object, that points to the other config_struct_body for the second device, and then do two calls as in the codegen solution.

erwango replied:
> That is not entirely true, there is a lot more parameters.

Indeed, my goal was not to oversimplify codegen solution. I should have used following syntax:
devicedeclare('compat_v1', 'config_struct_body_v1', 'data_struct_body', ....)

> But your use-case is totally supported

Ok, great removing my NACK. But tbh I'd prefer to have things available (even as a prototype) rather than doable and will be done if RFC is approved (even if I do trust you).

erwango reviewed (changes requested):
Thanks for all the changes, starting to looks nice.
2 remaining points

superna9999 reviewed (approved):
I'm ok for the i2c and spi stm32 drivers changes

galak commented:
@evenl I updated the topic-EDTS branch, can you rebase on it.

dbkinder reviewed (approved):
+1 for docs

evenl commented:
I was wondering if there is any progress regarding this topic? I think all questions/change requests has been addressed.

SebastianBoe commented:
@erwango : Hi, I see you are registered as a NACK. Can you re-asses your review?

SebastianBoe commented:
Hi, @galak if you could approve this RFC we can consider the RFC to be approved. We will then remove the RFC tag and rebase it onto the topic-EDTS branch as a PR.

SebastianBoe commented:
Hi, since there have been several approvals I am going to consider this RFC approved.

@evenl : You may now remove the RFC tag and begin preparing this so that it can be reviewed again as a PR and then merged.

erwango commented:
@SebastianBoe 

> .. and then merged.

Is that decided we're using jinja2 instead of codegen or other ? Have I missed something?

SebastianBoe commented:
> Is that decided we're using jinja2 instead of codegen or other ? Have I missed something?

Yes, that is what this RFC is for and what you have approved.
The implementation is still up for review.

erwango reviewed (commented):
Commenting to withdraw my approval, since approval equals "vote for" according to @SebastianBoe .

erwango reviewed (changes requested):
Trying to remove approval, or dismiss review..

SebastianBoe commented:
RFC is no longer considered to be approved due to misunderstanding.

Further review will be required ...

carlescufi commented:
>Is it really ? @carlescufi are you aligned on this?

Unfortunately the final decision on what templating system to use has been deferred to after the LTS release in order to get everything ready for the upcoming release.
Let's wait until then in order to settle the discussion.
